### PR TITLE
feat: add flat_size and fix to_binary

### PR DESCRIPTION
**Pull request by polvalente**

### CODE REVIEW

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -3295,6 +3298,9 @@ defmodule Nx do
   Returns the number of elements in the tensor.
 
   If a tuple is given, it returns the number of elements in a tensor with that shape.
+  Vectorized tensors will return the base shape's size.
+
+  See also: `flat_size/1`
```
Comment (josevalim) [inline nx/lib/nx.ex:3303]:
```suggestion
  Vectorized tensors will not include vectorized axes sizes. See `flat_size/1`.
```

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -3307,11 +3313,43 @@ defmodule Nx do
...
+      6
+
   """
   @doc type: :shape
   def size(shape) when is_tuple(shape), do: Tuple.product(shape)
   def size(tensor), do: size(shape(tensor))
 
+  @doc """
+  Returns the number of elements in the tensor.
+  Vectorized tensors will behave as if they aren't vectorized.
```
Comment (josevalim) [inline nx/lib/nx.ex:3326]:
```suggestion
  Returns the number of elements in the tensor (including vectorized axes).
```

File: nx/lib/nx.ex [hunk 3]
```diff
@@ -1721,21 +1721,24 @@ defmodule Nx do
       iex> Nx.to_binary(Nx.tensor([1.0, 2.0, 3.0]), limit: 2)
       <<1.0::float-32-native, 2.0::float-32-native>>
 
+      iex> Nx.to_binary(Nx.vectorize(Nx.tensor([1, 2, 3]), :x), limit: 2)
+      <<1::64-native, 2::64-native>>
+
+      iex> Nx.to_binary(Nx.vectorize(Nx.tensor([[1, 2, 3], [4, 5, 6]]), :x))
+      <<1::64-native, 2::64-native, 3::64-native, 4::64-native, 5::64-native, 6::64-native>>
```
[the remark this answers is not in the document]
Comment (josevalim) [reply nx/lib/nx.ex:1728]: To be clear, let's add them to the vectorized_test.exs.

**Review by josevalim — APPROVED** [review]

File: nx/lib/nx.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (josevalim) [reply nx/lib/nx.ex:1728]: So what if we add explicit sections? Such as:

```
### Vectorized tensors

`to_binary/1` returns the whole tensor, discarding vectorized axes:

    iex> ...
```

We probably want to add a mention to other functions that build on top of `to_binary`, such as `to_flat_list` and `to_list`.

[the remark this answers is not in the document]
Comment (polvalente) [reply nx/lib/nx.ex:1728]: Sounds good

File: nx/lib/nx.ex [hunk 4]
```diff
@@ -3307,11 +3340,42 @@ defmodule Nx do
...
+      iex> Nx.flat_size(Nx.vectorize(v1, :y))
+      24
+  """
+  @doc type: :shape
+  def flat_size(%T{vectorized_axes: axes} = tensor) when axes != [] do
+    base_size = size(tensor)
+    Enum.reduce(axes, base_size, fn {_, size}, acc -> acc * size end)
+  end
+
+  def flat_size(tensor), do: size(tensor)
```
Comment (josevalim) [inline nx/lib/nx.ex:3377]: Should we also introduce `flat_rank`? Anyway, it can also be done later. :)

**Review by josevalim — APPROVED** [review]
Beautiful.

**Pull request merged**